src/treesitter.rs

```diff
@@ -21,3 +25,103 @@ impl From<&'_ Node<'_>> for Match {
         Self::new(range.start_byte, range.end_byte)
     }
 }
+
+pub trait Parseable {
+    fn parse(&self, parser: &mut Parser, old_tree: Option<&Tree>) -> Option<Tree>;
+}
+
+impl<'a> Parseable for &'a [u8] {
+    fn parse(&self, parser: &mut Parser, old_tree: Option<&Tree>) -> Option<Tree> {
+        parser.parse(self, old_tree)
+    }
+}
+
+impl<'a> Parseable for &'a Rope {
+    fn parse(&self, parser: &mut Parser, old_tree: Option<&Tree>) -> Option<Tree> {
+        parser.parse_with(
+            &mut |byte_offset, _| {
+                let (chunk, chunk_start_byte_index, _, _) = self.chunk_at_byte(byte_offset);
+                &chunk[byte_offset - chunk_start_byte_index..]
+            },
+            old_tree,
+        )
+    }
+}
+
+#[derive(Copy, Clone)]
+pub enum RopeOrSlice<'a> {
+    Slice(&'a [u8]),
+    Rope(&'a Rope),
+}
+
+impl<'a> TextProvider<'a> for RopeOrSlice<'a> {
+    type I = RopeOrSliceTextProviderIterator<'a>;
+
+    fn text(&mut self, node: Node) -> Self::I {
+        match self {
+            Self::Slice(slice) => {
+                RopeOrSliceTextProviderIterator::Slice(iter::once(&slice[node.byte_range()]))
+            }
+            Self::Rope(rope) => {
+                let rope_slice = rope.byte_slice(node.byte_range());
+                RopeOrSliceTextProviderIterator::Rope(RopeOrSliceRopeTextProviderIterator::new(
+                    rope_slice,
+                    |rope_slice| rope_slice.chunks(),
```

Comment:
Ok so here I convinced myself that I needed a self-referential data structure

Why? Because we have to create the `rope_slice` `RopeSlice` "locally" (because we have to derive it "dynamically" from the `rope`) and then really all we need to store in our `Iterator` type is its `Chunks` iterator, but there's no eg `.into_chunks()` (that would take `self` ownership of that locally-created value), just `.chunks()` which takes `&self` by reference (and its (ie the `Chunks`'s) lifetime depends on that `&self` lifetime (because it's an iterator over references to pieces of that `&self`)), but that `&self` lifetime would go out of scope at the end of the enclosing `text()` method call because it's a local variable to that method call

So therefore the only way I could picture achieving it was to say "ok `rope_slice` you can tag along for the `Iterator` too even though we don't really "need" you, we just need to "keep you alive long enough""

Aka a self-referential struct because the `Chunks`'s lifetime refers to that `RopeSlice` which at that point is another field in the same struct

So I tried out I think `self_cell` but it seemed a little clunkier so then tried `ouroboros` and that seemed smoother

Here this is how they have you initialize the dependent field